### PR TITLE
Implement dynamic provisioner for AWS EBS

### PR DESCRIPTION
**Pull request by jsafrane**
**Do not merge until PR #14537 is merged!**

Depending on PR #14537, this is implementation of a dynamic provisioner for Amazon. It does not implement any QoS classes, it creates volumes with the default setting (=cheapest/slowest?).

Also, it attaches these tags to the volumes so the system admin can see what created the volume and for which PVC:
"kubernetes.io/created-for/PVC/Namespace"
"kubernetes.io/created-for/PVC/Name"

The first commit is cherry-picked from aforementioned PR and I'll remove it when the PR is merged.

### CODE REVIEW

**Comment by markturansky** [comment]
LGTM.

@justinsb mind reviewing?  Ignore the controller commit.

**Comment by markturansky** [comment]
needs rebase.

can you squash the last two?

**Comment by jsafrane** [comment]
Squashed and rebased

**Comment by justinsb** [comment]
This LGTM.  I do think that we should probably create gp2 volumes by default; as they are not much more expensive and are generally much faster: http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/EBSVolumeTypes.html

I'm not sure why I didn't do this in the first place - I guess because CreateVolume was originally just for tests and we were creating small sizes where arguably magnetic is the better deal.

But we can easily fix this later (and probably add support for different QOS classes)

File: pkg/controller/persistentvolume/types.go [hunk 1]
```diff
@@ -26,10 +26,25 @@ import (
...
 	// This is an experimental feature and likely to change in the future.
-	createdForKey = "volume.extensions.kubernetes.io/provisioned-for"
+	provisionedForKey = "volume.experimental.kubernetes.io/provisioned-for"
+
+	// Name of a tag attached to a real volume in cloud (e.g. AWS EBS or GCE PD)
+	// with namespace of a persistent volume claim used to create this volume.
+	cloudVolumeCreatedForNamespaceTag = "kubernetes.io/created-for/PVC/Namespace"
+	// Name of a tag attached to a real volume in cloud (e.g. AWS EBS or GCE PD)
+	// with name of a persistent volume claim used to create this volume.
+	cloudVolumeCreatedForNameTag = "kubernetes.io/created-for/PVC/Name"
```
Comment (liggitt) [inline pkg/controller/persistentvolume/types.go]: lowercase these?

Comment (markturansky) [inline pkg/controller/persistentvolume/types.go]: We can also tag with the pv name as well as the claim name:

```
kubernetes.io/created-for/pv-name
kubernetes.io/created-for/pvc-namespace
kubernetes.io/created-for/pvc-name
```

This ties the EBS volume (or whichever) to a PV, a namespace, and a PVClaim.

File: pkg/volume/aws_ebs/aws_ebs.go [hunk 2]
```diff
@@ -348,3 +389,67 @@ func (c *awsElasticBlockStoreCleaner) TearDownAt(dir string) error {
...
+}
+
+func (c *awsElasticBlockStoreProvisioner) NewPersistentVolumeTemplate() (*api.PersistentVolume, error) {
+	// Provide dummy api.PersistentVolume.Spec, it will be filled in
+	// awsElasticBlockStoreProvisioner.Provision()
+	return &api.PersistentVolume{
+		ObjectMeta: api.ObjectMeta{
+			GenerateName: "pv-aws-",
+			Labels: map[string]string{
+				"createdby": "aws_ebs_dynamic_provisioner",
```
Comment (markturansky) [inline pkg/volume/aws_ebs/aws_ebs.go]: Needs "kubernetes.io/" prefix.  Same for all provisioners.

We're not to use naked labels on an object.  We leave that empty namespace for the user.

File: pkg/cloudprovider/providers/aws/aws.go [hunk 3]
```diff
@@ -1216,15 +1217,15 @@ func (aws *AWSCloud) DetachDisk(instanceName string, diskName string) error {
...
 	// TODO: Should we tag this with the cluster id (so it gets deleted when the cluster does?)
 	// This is only used for testing right now
 
 	request := &ec2.CreateVolumeInput{}
-	request.AvailabilityZone = &aws.availabilityZone
+	request.AvailabilityZone = &s.availabilityZone
 	volSize := (int64(volumeOptions.CapacityMB) + 1023) / 1024
 	request.Size = &volSize
-	response, err := aws.ec2.CreateVolume(request)
+	response, err := s.ec2.CreateVolume(request)
```
Comment (markturansky) [inline pkg/cloudprovider/providers/aws/aws.go]: What EBS VolumeType is this by default?

We have a request for "gp2 SSD".

File: pkg/volume/aws_ebs/aws_ebs.go (hunk 2 above)
Comment (jsafrane) [inline pkg/volume/aws_ebs/aws_ebs.go]: Done, also moved from `Labels` to `Annotations`, following HostPath example.

File: pkg/cloudprovider/providers/aws/aws.go (hunk 3 above)
Comment (jsafrane) [inline pkg/cloudprovider/providers/aws/aws.go]: AWS defaults to 'standard', i.e. Magnetic.

File: pkg/controller/persistentvolume/types.go (hunk 1 above)
Comment (jsafrane) [inline pkg/controller/persistentvolume/types.go]: swithed to lowercase, `kubernetes.io/created-for/pvc/namespace` and `kubernetes.io/created-for/pvc/name`

File: pkg/cloudprovider/providers/aws/aws.go (hunk 3 above)
Comment (markturansky) [inline pkg/cloudprovider/providers/aws/aws.go]: Can you add some bits to make the default for the provisioner ssd?

Comment (jsafrane) [inline pkg/cloudprovider/providers/aws/aws.go]: Done, in an extra commit.

**Comment by jsafrane** [comment]
Rebased to latest version of #14537 and tested on AWS.

**Comment by jsafrane** [comment]
@k8s-bot test this

**Comment by markturansky** [comment]
@saad-ali this next, please?

We'll need to rebase this after #14537 is merged.

**Comment by saad-ali** [comment]
Taking a look. Will try to tease apart the changes from 14537

File: pkg/volume/aws_ebs/aws_util.go [hunk 4]
```diff
@@ -107,3 +108,42 @@ func (util *AWSDiskUtil) DetachDisk(c *awsElasticBlockStoreCleaner) error {
 	}
 	return nil
 }
+
+func (util *AWSDiskUtil) DeleteVolume(d *awsElasticBlockStoreDeleter) error {
+	cloud := d.plugin.host.GetCloudProvider()
```
Comment (saad-ali) [inline pkg/volume/aws_ebs/aws_util.go]: nit: this code to fetch and cast the cloud provider gets repeated a lot, factor it out into a util method, if possible.

File: pkg/volume/aws_ebs/aws_util.go [hunk 5]
```diff
@@ -107,3 +108,42 @@ func (util *AWSDiskUtil) DetachDisk(c *awsElasticBlockStoreCleaner) error {
...
+func (util *AWSDiskUtil) CreateVolume(c *awsElasticBlockStoreProvisioner) (string, int, error) {
+	const mega = 1024 * 1024
+	cloud := c.plugin.host.GetCloudProvider()
+	if cloud == nil {
+		glog.Errorf("Cloud provider not initialized properly")
+		return "", 0, errors.New("Cloud provider not initialized properly")
+	}
+
+	// convert to MB with rounding
+	volSize := int((c.options.Capacity.Value() + mega - 1) / mega)
```
Comment (saad-ali) [inline pkg/volume/aws_ebs/aws_util.go]: Could you add a comment on how this work.

File: pkg/volume/aws_ebs/aws_ebs.go [hunk 6]
```diff
@@ -349,3 +390,67 @@ func (c *awsElasticBlockStoreCleaner) TearDownAt(dir string) error {
...
+var _ volume.Provisioner = &awsElasticBlockStoreProvisioner{}
+
+func (c *awsElasticBlockStoreProvisioner) Provision(pv *api.PersistentVolume) error {
+	volumeID, size, err := c.manager.CreateVolume(c)
+	if err != nil {
+		return err
+	}
+	pv.Spec.PersistentVolumeSource.AWSElasticBlockStore.VolumeID = volumeID
+	pv.Spec.Capacity = api.ResourceList{
+		api.ResourceName(api.ResourceStorage): resource.MustParse(fmt.Sprintf("%d", size)),
```
Comment (saad-ali) [inline pkg/volume/aws_ebs/aws_ebs.go]: No units suffix? Are they assumed?

**Comment by saad-ali** [comment]
Overall looks good, some minor comments (and needs rebase).

**Comment by saad-ali** [comment]
Ignoring this PR, since it appears to be replaced by #18601

**Comment by jsafrane** [comment]
Closing in favor of https://github.com/kubernetes/kubernetes/pull/18601

**Pull request closed without merging**